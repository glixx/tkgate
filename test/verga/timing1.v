module top;
  integer i;
   reg 	  ck;
  reg data,en;
  wire out;

  initial
    $monitor("%4d: ck=%b data=%b en=%b - out=%b",$time,ck,data,en,out);

  always
    #50 ck = !ck;

  initial
    begin
      ck = 1'b1;
      data = 1'b0; en = 1'b1;


      #95;
      data = 1'b1; en = 1'b1;
      #5;
      $display("Expect $setup violation at time %d.",$time); 

      #85;
      data = 1'b0; en = 1'b1;
      #15;
      $display("Expect no violations at time %d.",$time); 

      #91; 
      data = 1'b1; en = 1'b1;
      #9;
      $display("Expect $setup violation at time %d.",$time); 

      #50; 
      data = 1'b0; en = 1'b1;
      #50;
      $display("Expect no violations at time %d.",$time); 

      #95; 
      data = 1'b1; en = 1'b0;
      #5;
      $display("Expect no violations at time %d.",$time); 

      #2; 
      $display("Expect no violations at time %d.",$time); 
      data = 1'b1; en = 1'b1;
      #98;

      #2;
      $display("Expect $hold violation at time %d.",$time); 
      data = 1'b0; en = 1'b1;
      #98;

      #11;
      $display("Expect no violation at time %d.",$time); 
      data = 1'b1; en = 1'b1;
      #89;

      $finish;	
    end

  latch l1(ck,data,en,out);

endmodule

module latch(ck,data,en,out);
input ck,data,en;
output out;
reg out = 1'bx;

  specify
    $setup(data, posedge ck &&& en, 10);
    $hold(posedge ck &&& en, data, 10);
  endspecify

  always @(posedge ck)
    if (en)
      begin
	 out = data;
      end

endmodule
