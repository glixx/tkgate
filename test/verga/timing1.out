   0: ck=1 data=0 en=1 - out=0
  50: ck=0 data=0 en=1 - out=0
  95: ck=0 data=1 en=1 - out=0
Expect $setup violation at time 100.
Runtime Error at 100: Timing violation in top.l1[latch] $setup(data, (posedge ck) &&& en, 10).
 100: ck=1 data=1 en=1 - out=1
 150: ck=0 data=1 en=1 - out=1
 185: ck=0 data=0 en=1 - out=1
Expect no violations at time 200.
 200: ck=1 data=0 en=1 - out=0
 250: ck=0 data=0 en=1 - out=0
 291: ck=0 data=1 en=1 - out=0
Expect $setup violation at time 300.
Runtime Error at 300: Timing violation in top.l1[latch] $setup(data, (posedge ck) &&& en, 10).
 300: ck=1 data=1 en=1 - out=1
 350: ck=0 data=0 en=1 - out=1
Expect no violations at time 400.
 400: ck=1 data=0 en=1 - out=0
 450: ck=0 data=0 en=1 - out=0
 495: ck=0 data=1 en=0 - out=0
Expect no violations at time 500.
 500: ck=1 data=1 en=0 - out=0
Expect no violations at time 502.
 502: ck=1 data=1 en=1 - out=0
 550: ck=0 data=1 en=1 - out=0
 600: ck=1 data=1 en=1 - out=1
Expect $hold violation at time 602.
Runtime Error at 602: Timing violation in top.l1[latch] $hold((posedge ck) &&& en, data, 10).
 602: ck=1 data=0 en=1 - out=1
 650: ck=0 data=0 en=1 - out=1
 700: ck=1 data=0 en=1 - out=0
Expect no violation at time 711.
 711: ck=1 data=1 en=1 - out=0
 750: ck=0 data=1 en=1 - out=0
